### ISSUE TITLE
application crashes if runtime is missing

### ISSUE THREAD

**Issue by tvogel-nid** [AaronRobinsonMSFT/DNNE#79]
We have currently the problem that a GUI app (LabVIEW, we have no control of) that is calling into the generated *.NE.dll crashes without any error.
So I'm looking for a way to prevent the call into abort() in case of a missing runtime.
I was hoping for the mentioned methods preload_runtime() and/or set_failure_callback() but I somehow miss whether these could be used for issuing an error to the GUI and not crashing it.
Could these function be the right way to go and if yes, are there any examples available I could use as guidance?
If not, is there any other way to achieve this?

Thanks a lot Thomas

**Comment by AaronRobinsonMSFT** [comment]
@tvogel-nid There is presently no way to address this since I'm not sure how to handle the case in practice. Let me walk through the scenario I have in mind.

A managed binary exposes the following function:

```csharp
IntPtr DoWork(int a, float f);
```

The native export looks like the following:

```c
void* DoWork(int a, float f);
```

The question I am struggle with is, what happens if the runtime fails to load? Do I return `NULL`? That could indicate something completely unexpected and likely not what is desired.

I suppose I could provide another option, attribute? ,for users to define "default" values for each method in the event the CLR fails to load but I'm not really a fan of that in practice. I guess we could also provide a native callback that could be supplied that indicates "give me data of this size because the runtime has failed to load" &ndash; that is rather complicated too.

My initial thinking here was that the CLR is the equivalent of a native dependency so the failure should be the same. Do you have a suggested UX for this type of failure?

**Comment by tvogel-nid** [comment]
@AaronRobinsonMSFT even after thinking it over and over I still have no real idea how to overcome this problem.
Would it be an option to call just exit() instead of abort()? Maybe selectable by a compile switch.
Problem seems to be that my controlling application(LabVIEW) uses so called "call library function nodes" to call into the DLL and reacts with a full crash without any warning/error to a SIGABRT that is issued by the abort() as far as I understand.

**Comment by AaronRobinsonMSFT** [comment]
@tvogel-nid I can absolutely change how the application fails. In fact, I can also permit users to provide their own native code during a runtime load or export discovery failure. This would involve consumers providing C code &ndash; I assume this is okay.

I can think of two options:

1) Create a new .NET `DNNE` Attribute that takes a snippet of C code that will be compiled into the final binary for these failure cases.

1) Have an MSBuild `Item` that permits setting additional C files that will be included in the native export binary. The native API would be expanded to allow registration at run-time or perhaps during link where the user defines their own failure callbacks in C.

Both approaches wouldn't change the current behavior of the `DNNE` fallback being `abort` since the user defined native callbacks could call `exit` or whatever they like instead. However, if the user still didn't define an process termination strategy the system would failure as currently defined.

Do you have a preference here? Other takes?

**Comment by AaronRobinsonMSFT** [comment]
@tvogel-nid I am going to follow @jkoritzinsky's suggestion and permit an override mechanism. This will enable you or anyone else to override the default behavior as desired. Once could imagine a Dialog box being presented that would indicate to the user what to install and once the user acknowledged that the process would terminate via any number of ways&mdash;`exit()`, `abort()`, etc. This permits people the ability to provide a nice reason as to why something is going down.

If this is something needed soon, let me know and I will publish a new package otherwise I'll wait a bit to see if anyone has other needs.

**Comment by AaronRobinsonMSFT** [comment]
This support was published in 1.0.28 &ndash; https://www.nuget.org/packages/DNNE/1.0.28.
